### ISSUE TITLE
Using StencilFlow with Python source code

### ISSUE THREAD

**Issue by gabrielrodcanal** [spcl/stencilflow#6]
Hi there,
I am interested in optimising a stencil code written in Python using StencilFlow. From what I see in the repository and from what I read on your paper it is possible to do so by generating the SDFG from DACE and then transforming this into a JSON StencilFlow file with the tool `bin/sdfg_to_stencilflow.py`. Unfortunately, I haven't been able to do so, as I'm getting the error `ValueError: Program has no outputs`. Here a small code to reproduce:

```
import dace
import numpy as np

DIM = 10

@dace.program
def jacobi(a: dace.float64[DIM,DIM], b: dace.float64[DIM,DIM]):
    for i in range(1,DIM-1):
        for j in range(1,DIM-1):
            b[i] = b[i] + 2*a[-1,0] - 2*a[1,0] + 2*a[0,-1] - 2*a[0,1]

if __name__ == "__main__":
    a = np.array([[0 for j in range(DIM)] for i in range(DIM)], dtype=np.double)
    b = np.array([[0 for j in range(DIM)] for i in range(DIM)], dtype=np.double)
    jacobi(a,b)
```

I just fed the `.sdfg` file in `.dacecache/jacobi` to `bin/sdfg_to_stencilflow.py`. Is there any fundamental annotation that's missing? I wouldn't want to modify the source code, is it possible to run StencilFlow on this code as-is other than some basic annotations?

Note: I am using the DACE version that is frozen as submodule in this repository.

Thanks,
Gabriel

**Comment by definelicht** [comment]
Hey Gabriel, I can try reproducing myself when I'm back a computer, probably the stencil detection in DaCe programs is somewhat sensitive to the format.

Did you consider writing the JSON directly? It should be fairly straightforward if you program is not humongous 🙂

**Comment by gabrielrodcanal** [comment]
Hey Johannes,
Yeah, I've seen the JSON is quite nice and straightforward, but my program is actually quite large. If you could try to reproduce the issue and let me know if what I'm trying to do is currently supported that'd be helpful :).

**Comment by definelicht** [comment]
I apologize for the wait, I realize it might be too late for you now. I am no longer in academia, so it's hard to find some free time to look into this, and I no longer have access to a machine with an FPGA or with the FPGA tools installed.

It took me a while to get DaCe running, it seems to only work with Python 3.8 and Sympy 1.7, but I finally managed to reproduce your issue.

It's been a long time since I worked with this code, but looking at the code I don't believe `sdfg_to_stencilflow.py` is intended to infer/detect stencils; it relies on stencil library nodes already being present in the program. You would need to provide an input SDFG that is already in this form. There is some stencil detection functionality in DaCe, but I think it's risky to rely on this in the general case.

Unfortunately I think the only solution that would reliably give you a representative program with StencilFlow is generating the input JSON files, if not by hand, then somehow scripted.
Keep in mind that the focus of the paper is compiling a stencil DSL; not so much inferring stencils from imperative code in the first place - the code you found is fairly specialized bridge to handle specific stencil programs output by a climate/weather frontend.
